consult-yasnippet.el

```diff
@@ -0,0 +1,158 @@
+  "Expand the yasnippet template TEMPLATE at point."
+  (deactivate-mark)
+  (goto-char (car consult-yasnippet--region))
+  (when (not (string-equal "" consult-yasnippet--region-contents))
+    (push-mark (point))
+    (push-mark (cdr consult-yasnippet--region) nil t))
+
+  (yas-expand-snippet (yas--template-content template)
+                      nil nil
+                      (yas--template-expand-env template)))
+
+(defun consult-yasnippet--preview (template _)
+  "Previewer for `consult--read'.
+This function expands TEMPLATE at point in the buffer
+`consult-yasnippet--read-snippet' was started in. This includes
+overwriting any region that was active and removing any previous
+previews that're already active.
+
+When TEMPLATE is not given, this function essentially just resets
+the state of the current buffer to before any snippets were previewed."
+  (with-current-buffer consult-yasnippet--buffer
+    (let ((yas-verbosity 0)
+          (inhibit-redisplay t)
+          (inhibit-read-only t)
+          (orig-offset (- (point-max) (cdr consult-yasnippet--region)))
+          (yas-prompt-functions '(yas-no-prompt)))
+
+      ;; We always undo any snippet previews before maybe setting up
+      ;; some new previews.
+      (delete-region (car consult-yasnippet--region)
+                     (cdr consult-yasnippet--region))
+      (goto-char (car consult-yasnippet--region))
+      (setcar consult-yasnippet--region (point))
+      (insert consult-yasnippet--region-contents)
+      (setcdr consult-yasnippet--region (point))
+
+      (when template
+        (unwind-protect
+            (consult-yasnippet--expand-template template)
+          (unwind-protect
+              (mapc #'yas--commit-snippet
+                    (yas-active-snippets (point-min) (point-max)))
+            (setcdr consult-yasnippet--region (- (point-max) orig-offset))))
+        (redisplay)))))
+
+(defmacro consult-yasnippet--setup (&rest body)
+  "Setup the local variables and environment for `consult-yasnippet'.
+This environment is used both in `consult-yasnippet--preview'and
+`consult-yasnippet--expand-template'."
+  `(progn
+     (barf-if-buffer-read-only)
+     (unless (bound-and-true-p yas-minor-mode)
+       (error "`yas-minor-mode' not enabled in current buffer"))
+
+     (let* ((consult-yasnippet--snippets
+             (mapcar (lambda (template)
+                       (cons (yas--template-name template) template))
+                     (yas--all-templates (yas--get-snippet-tables))))
+            (consult-yasnippet--buffer (current-buffer))
+            (consult-yasnippet--region (if (region-active-p)
+                                           (cons (region-beginning) (region-end))
+                                         (cons (point) (point))))
+            (consult-yasnippet--region-contents (buffer-substring (car consult-yasnippet--region)
+                                                                  (cdr consult-yasnippet--region))))
+       ,@body)))
```

Comment:
Most of this should probably go to the preview function then when it captures the state.

Reply:
Done.